pkg/client/interceptor/interceptor.go

```diff
@@ -4,3 +4,191 @@
 
 // Package interceptor contains gRPC client interceptors related to authentication and authorization.
 package interceptor
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+
+	"github.com/siderolabs/go-api-signature/pkg/message"
+)
+
+// SkipInterceptorContextKey is a context key used to skip interceptor to avoid infinite recursion.
+type SkipInterceptorContextKey struct{}
+
+// SignerFunc is a function that is called to get a signer.
+//
+// If the returned bool is false, the interceptor will not attempt to re-authenticate and retry the call.
+type SignerFunc func(ctx context.Context, cc *grpc.ClientConn) (signer message.Signer, retryable bool, err error)
```

Comment:
`retryable bool` - now when we read a signer, we know if we can attempt to renew it (e.g. when it is a service account, we do not retry).